### PR TITLE
kops-controller using the cluster-api

### PR DESCRIPTION
**Pull request by justinsb**
This can map an InstanceGroup to a MachineDeployment; currently only
implemented for GCE, but working there on my integration branch.

Needs a number of supporting fixes, for example the GCE alias ip range
PRs for kops, and also a bunch of PRs to cluster-api-provider-gcp
(which are pending).

### CODE REVIEW

File: cmd/kops-controller/pkg/controller/controller.go [hunk 1]
```diff
@@ -14,34 +13,21 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package main
+package controller
```
Comment (mikesplain) [inline cmd/kops-controller/pkg/controller/controller.go]: I was going to say, I thought this code looked familiar, great to see it being reused!

**Comment by chrisz100** [comment]
Excited to see this in upstream, too!

**Comment by Sn0rt** [comment]
a great job.

Is it that kops will become a cluster-api front end ?

**Comment by mitch000001** [comment]
I really appreciate this PR. What it enables is just great. I would however try a different approach within `cmd/kops-controller/pkg/clusterapi/gce.go`, as the current one can lead to a lot of if/else branches within that file, specifically the `BuildMachineDeployment` function. In my opinion there are two options. Either we can extract a common interface for all cloud providers or we rename the builder to be specific to GCE to reflect that it is only scoped for that at the moment.
I know that we should not generalize prematurely, but having a package with the generic name `clusterapi` make we wonder if we want to bury all cloud implementations within that.

**Comment by rifelpet** [comment]
I believe we can close this out as it was split into #7496 (merged) and #8151 but if I'm wrong feel free to reopen.

/close

**Pull request closed without merging**